### PR TITLE
gcc.libgcc: compare host and target platforms, rather than their triples

### LINKED ISSUE #263040: Build failure: `gcc` fails to compile on linux when nixpkgs is configured to cross-compile back to the same system and `libcCross` does not contain `passthru.libgcc`
GCC fails to compile when nixpkgs is configured to cross-compile back to the same system and `libcCross` does not contain `passthru.libgcc`. Most of the GCC build determines whether it's cross-compiling by testing if `targetPlatform != hostPlatform`, but `preFixupLibGccPhase` instead tests `hostPlatform.config != targetPlatform.config`. This means that when building under these circumstances it does most of the build and then fails to find `libgcc_s.so` as it's looking in the wrong place. The `preFixupLibGccPhase` code needs to use the same test as the code that defines `targetConfig`.

This isn't normally noticed as `glibcCross` includes `passthru.libgcc`, but if I use `crossOverlays` to replace it with a libc that does not have this passthru, that's when GCC breaks. This is blocking me as I am attempting to backdate glibc to an older version that doesn't bundle libgcc (for symbol versioning reasons).

`targetConfig` definition:
https://github.com/NixOS/nixpkgs/blob/38d33bc5fafd56885b33891e86e9305cedffd89f/pkgs/development/compilers/gcc/default.nix#L356

`preFixupLibGccPhase` tests:
https://github.com/NixOS/nixpkgs/blob/ca03f204a225445f9497b2e77161abee5aa10fe0/pkgs/development/compilers/gcc/common/libgcc.nix#L38-L41
https://github.com/NixOS/nixpkgs/blob/ca03f204a225445f9497b2e77161abee5aa10fe0/pkgs/development/compilers/gcc/common/libgcc.nix#L86-L89

### Steps To Reproduce

```bash
nix build --impure --expr 'let pkgs = import (builtins.getFlake "nixpkgs/5ba549eafcf3e33405e5f66decd1a72356632b96") {
  crossOverlays = [(self: super: {
    glibcCross = super.glibcCross.overrideAttrs (old: {
      passthru = builtins.removeAttrs old.passthru [ "libgcc" ];
    });
  })];
}; in pkgs.pkgsLLVM.buildPackages.gcc.cc'
```

### Build log

Relevant output from the end of the log:

```
Moving /nix/store/qmpisiz34xp8757sj3dbrxqv2ha394jk-x86_64-unknown-linux-gnu-stage-final-gcc-12.3.0/x86_64-unknown-linux-gnu/lib/libgcc_s.so.1 to /nix/store/1d1dgmqfk54cdg7kr3rz7761qrcldifp-x86_64-unknown-linux-gnu-stage-final-gcc-12.3.0-lib/x86_64-unknown-linux-gnu/lib/libgcc_s.so.1
[...]
preFixupLibGccPhase
mv: cannot stat '/nix/store/1d1dgmqfk54cdg7kr3rz7761qrcldifp-x86_64-unknown-linux-gnu-stage-final-gcc-12.3.0-lib/lib/libgcc_s.so': No such file or directory
```

Notice now the `libgcc_s.so.1` path includes the target config in the lib path, but the `preFixupLibGccPhase` tries to look for it without the target config.

### Additional context

If I override the package to set `targetConfig` to `null` it seems to skip building `libgcc_s.so` entirely, so it looks like the `targetConfig` definition is correct and the `libgcc.nix` code needs to be updated to use the same test (or possibly to just check if `targetConfig != null` instead of trying to independently test for cross-compilation).

### Notify maintainers

@Synthetica9 @vcunat @Ericson2314

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix run nixpkgs#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.51, NixOS, 23.11 (Tapir), 23.11.20230919.5ba549e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

The rest of our gcc expression prepends `"${targetPlatform.config}-"`    to paths and binaries if `hostPlatform!=targetPlatform`.  The    `libgcc.nix` expression was using `hostPlatform.config!=targetPlatform.config`, which caused it to    look in the wrong place when moving files.  This commit corrects that.

Closes #263040 

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build pkgsCross.gnu64_simplekernel.bash

File: pkgs/shells/bash/5.nix [hunk 1]
```diff
@@ -91,7 +91,7 @@ stdenv.mkDerivation rec {
 
   strictDeps = true;
   # Note: Bison is needed because the patches above modify parse.y.
-  depsBuildBuild = [ buildPackages.stdenv.cc ];
+  depsBuildBuild = [ stdenv ];
```
Comment (Ericson2314) [inline pkgs/shells/bash/5.nix:94]: Why this change? I've never seen depending on entire other stdenvs like that.

Comment (Ericson2314) [reply pkgs/shells/bash/5.nix:94]: If we just skip this, everything else looks good to me.

Comment (amjoseph-nixpkgs) [reply pkgs/shells/bash/5.nix:94]: > Why this change?

Because `!(buildPackages.stdenv.cc?__spliced)`.

See https://github.com/NixOS/nixpkgs/pull/264992 for the test case.

The reproducer requires two platforms which have the same triple but aren't `==`-equal to each other.  People are creating them on their own and getting burned by this.  Latest example:

- https://github.com/NixOS/nixpkgs/issues/265121

There aren't any in `lib.systems.examples`, so #264992 adds one in order to be able to answer your question "why this change".

Comment (amjoseph-nixpkgs) [reply pkgs/shells/bash/5.nix:94]: > Why this change?

Because: https://github.com/NixOS/nixpkgs/pull/264992#issuecomment-1791961285

Comment (amjoseph-nixpkgs) [reply pkgs/shells/bash/5.nix]: > If we just skip this, everything else looks good to me.

I moved it into

- https://github.com/NixOS/nixpkgs/pull/265162

**Comment by lilyball** [comment]
It looks like this should fix https://github.com/NixOS/nixpkgs/issues/263040, except you missed a comparison, `libgcc.nix` does another config check in `preFixupLibGccPhase` that needs to be updated.

**Comment by amjoseph-nixpkgs** [comment]
> It looks like this should fix #263040, except you missed a comparison, `libgcc.nix` does another config check in `preFixupLibGccPhase` that needs to be updated.

Thank you @lilyball!  Verifying this now.

**Comment by amjoseph-nixpkgs** [comment]
> Thank you @lilyball! Verifying this now.

I have verified that your test case (deflakified below) builds correctly at a37a7d2f1293ec3845717e388d12b9395b3efbd9

```
nix build --impure --expr 'let pkgs = import ./. {
  crossOverlays = [(self: super: {
    glibcCross = super.glibcCross.overrideAttrs (old: {
      passthru = builtins.removeAttrs old.passthru [ "libgcc" ];
    });
  })];
}; in pkgs.pkgsLLVM.buildPackages.gcc.cc'
```

**Comment by amjoseph-nixpkgs** [comment]
I rebuilt the entire `pkgs.tests.cross.sanity` mega-regression-suite at commit a37a7d2f1293ec3845717e388d12b9395b3efbd9.

This won't break anything.  If it does I take full responsibility.
